---
title: "2023-05-23_Data_RNAseq"
output:
  html_document:
    df_print: paged
date: "2023-05-23"
runtime: html
editor_options:
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load Libraries

These libraries are required to perform this RNAseq analysis. Make sure you have install them before begin the analysis.

```{r}
suppressPackageStartupMessages({
library(tidyverse)
library(tidybulk)
library(janitor)
library(plotly)
library(ggrepel)
library(DESeq2)
library(GeneTonic)
library(pcaExplorer)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(topGO)
})
```

# Data Dictionary

Each sample was organized into its owned folder named using various variable separated by underscore ("\_"). For example, for"697_C\_CAR_Rep2_S5_L001":

-   697: cell line
-   C: treatment
-   CAR: condition
-   Rep2: replicate
-   S5: sequencing cell segment
-   L001: sequencing cell line

The code below manually listing samples' folder in a `vector` variable called `dir.names`

```{r}
dir.names = c(
"quant_files/D1_B7C2_Il8_stim_S8_L002",
"quant_files/D1_B7C2_bead_stim_S14_L002",
"quant_files/D1_B7C2_no_stim_S2_L002",
"quant_files/D1_B7_Il8_stim_S7_L002",
"quant_files/D1_B7_bead_stim_S13_L002",
"quant_files/D1_B7_no_stim_S1_L002",
"quant_files/D2_B7C2_Il8_stim_S10_L002",
"quant_files/D2_B7C2_bead_stim_S16_L002",
"quant_files/D2_B7C2_no_stim_S4_L002",
"quant_files/D2_B7_Il8_stim_S9_L002",
"quant_files/D2_B7_bead_stim_S15_L002",
"quant_files/D2_B7_no_stim_S3_L002",
"quant_files/D3_B7C2_Il8_stim_S12_L002",
"quant_files/D3_B7C2_bead_stim_S18_L002",
"quant_files/D3_B7C2_no_stim_S6_L002",
"quant_files/D3_B7_Il8_stim_S11_L002",
"quant_files/D3_B7_bead_stim_S17_L002",
"quant_files/D3_B7_no_stim_S5_L002"
)
```

A vector

```{r}
mode(dir.names)
str(dir.names)
```

Convert the `vector` to `data.frame`.

```{r}
Data.dictionary = data.frame(Orig.Name = dir.names, Folder.Name = dir.names) %>% 
  mutate(Folder.Name = str_remove(Folder.Name, pattern = "quant_files/"))
```

A `DataFrame`:

```{r}
Data.dictionary
```

## Transform Dictionary

Perform these steps to transform the `data.frame` into desirable format.

```{r}
Data.dictionary %>% 
  ## Create or duplicate the "Orig.Name" column and called it "Dup"
  mutate(Dup = Folder.Name) %>% 
  ## Breakup (separate) the "Dup" column using "_" as its separator
  separate_wider_delim(Dup, names = c("donor", "CAR", "stim", "line1", "line2", "line3"), delim = "_") %>%
  ## Create a new column called "col.name" by modifying "Orig.Name"
  ## by removing "_S(any digit)_L001"
  mutate(col.name = str_remove(Folder.Name, pattern = "_stim_S([:digit:]+)_L002")) %>%
  # Relocate the new "col.name" column right after the first column; "Org.Name"
  # relocate(col.name, .after = Orig.Name) %>% 
  ## Make "treatment" column a factor
  mutate(donor = as.factor(donor)) %>% 
  mutate(CAR = as.factor(CAR)) %>% 
  mutate(stim = as.factor(stim)) %>% 
  # mutate(col.name = str_c(donor, CAR, stim, sep = "_")) %>% 
  ## Remove the two columns "line1" and "line2", as we don't need them.
  dplyr::select(-c("line1", "line2", "line3")) -> Data.dictionary
```

```{r}
Data.dictionary
```

# Transcript-Gene-Pair: `tx2gene_df`

Our goal is to sum up all counts of a gene (represented by gene ID) from all its transcripts (represented by transcript ID). First, we will import gene's feature file (GTF) which contains gene id and transcipt id match.

```{r}
# Load this below
gtf = rtracklayer::import("data/Homo_sapiens.GRCh38.106.gtf")
## Transform into data frame for easy wrangling
gtf_df = as.data.frame(gtf)
gtf_df
```

### Save `gtf`

```{r}
save(gtf_df, file = "data/gtf_df.Rdata")
```

## Load the saved copy

```{r}
load(file = "data/gtf_df.Rdata")
```

## Look up table

Create a look-up-table matching gene id to all its transcript ids:

```{r}
gtf_df %>%
  dplyr::select(transcript_id, gene_id) %>%
	distinct(transcript_id, gene_id) %>%
  drop_na() -> tx2gene_df
tx2gene_df

```

# Salmon loop

Predefine count table matrix by importing one of Salmon result to determine the number of genes for the table matrix.

```{r}
## Assesss how many genes in each sample?
tmp = read.delim(paste("quant_files", Data.dictionary$Folder.Name[1], "quant.sf", sep = "/")) %>%
  ## The Ensembl ID (Name) column has an extra period (.) and a digit
  ## Remove them from each entries:
  mutate(Name = str_sub(Name, end=-3)) %>%
  ## Add gene id onto the table matching against transcript id:
  inner_join(tx2gene_df, by=c("Name"="transcript_id")) %>%
  ## Sum up all the counts for each gene id from all its transcript ids
  group_by(gene_id) %>%
  summarise(NumReads=as.integer(sum(NumReads)))

```

In this dataset, there are 35,113 genes interrogated.

```{r}
tmp
```

Now we can (pre)define the data matrix to hold all the samples, and start looping in all the samples from Salmon.

```{r}

sf.df = data.frame(matrix(0, nrow = length(tmp$gene_id), ncol = length(Data.dictionary$Orig.Name)))
colnames(sf.df) = Data.dictionary$col.name
rownames(sf.df) = tmp$gene_id
for(i in seq_along(Data.dictionary$Orig.Name)){
  cat("Reading ", Data.dictionary$Orig.Name[i], " Now ..............\n")

  sf.df[,i] = read.delim(paste("quant_files", Data.dictionary$Folder.Name[i], "quant.sf", sep = "/"))%>%
    mutate(Name = str_sub(Name, end=-3)) %>%
    inner_join(tx2gene_df, by=c("Name"="transcript_id")) %>%
    group_by(gene_id) %>%
    summarise(NumReads=as.integer(sum(NumReads))) %>%
    pull(NumReads)

}

```

A quick peek of the imported dataset shows that the data matrix has:

-   35,113 genes (rows)
-   18 samples (columns)

```{r}
sf.df
```

Save a copy of the Salmon dataset so you don't have to reimport all the files everytimes:

```{r}
save(sf.df, file = "data/sf.df.Rdata")
```

```{r}
load(file = "data/sf.df.Rdata")
```

# Define subset

```{r}
Data.dictionary
```

```{r}
Data.dictionary$col.name
```

```{r}
subset.samples = Data.dictionary$col.name
```

## Remove Outlier

```{r}
#subset.samples = 
```

## (Re)define meta data

```{r}
Data.dictionary %>% 
  dplyr::filter(col.name %in% subset.samples) -> Data.dictionary.final
```

# Subset and filter count data

```{r}
sf.df.data = sf.df %>%
  ## Select or subset the desired samples 
  #dplyr::select(subset.samples) %>%
  ## Remove all NA (if any)
  drop_na() %>% 
  ## Filter low expressors
  ### each value must be more than 10
  mutate(Rowsum = apply(., 1, function(x) sum(x > 10))) %>% 
  ### All values must be more than 10 (set to x if only x value is satisfied)
  dplyr::filter(Rowsum == dim(Data.dictionary.final)[1]) %>% 
  ## No longer need the column "Rowsum", we can remove it
  dplyr::select(-Rowsum)
sf.df.data
```

# DESeq Object

```{r}
class(sf.df.data)
```

## Create SummarizeExperiment Object

```{r}
sf.df.se = SummarizedExperiment(assays = list(counts = as.matrix(sf.df.data)),
                                colData = Data.dictionary.final)
```

## Create DESeq Object

```{r}
Data.dictionary.final
```

```{r}
sf.df.dds = DESeqDataSet(sf.df.se, design = ~CAR)
```

# PCA Plot

```{r}
sf.df.se %>% 
  keep_abundant() %>% 
  scale_abundance() %>% 
  reduce_dimensions(method = "PCA", .dims = 3) -> sf.df.se.pca
```

## 2D

```{r}
# Use colourblind-friendly colours
friendly_cols <- dittoSeq::dittoColors()

# Set theme
custom_theme <-
  list(
    scale_fill_manual(values = friendly_cols),
    scale_color_manual(values = friendly_cols),
    theme_bw() +
      theme(
        panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.major = element_line(size = 0.2),
        panel.grid.minor = element_line(size = 0.1),
        text = element_text(size = 9),
        legend.position = "bottom",
        strip.background = element_blank(),
        axis.title.x = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.title.y = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1)
      )
  )

```

```{r}
sf.df.se.pca %>% 
  pivot_sample() %>% 
  ggplot(aes(x = PC1, y = PC2, colour = donor)) +
  geom_point() +
  geom_text_repel(aes(label = .sample), show.legend = FALSE, size = 3) +
  custom_theme

```


## 3D

```{r}
colData(sf.df.se.pca)$col.name
```

```{r}
PCA.3D = sf.df.se.pca %>%
  pivot_sample() %>%
  plot_ly(
    x = ~`PC1`,
    y = ~`PC2`,
    z = ~`PC3`,
    color = ~stim,
    text = colData(sf.df.se.pca)$col.name
  )

htmlwidgets::saveWidget(
  widget = PCA.3D,
  file = "Data.PCA.html",
  selfcontained = T,
  # embed-resources = T,
  title = "Data PCA"
)

```

# DE Analysis

## Statistical Analysis

```{r}
sf.df.dds = DESeq(sf.df.dds)
```

## Analysis Result

```{r}
Data.dictionary.final
```


```{r}
res.sf.df.dds = results(sf.df.dds,
                        contrast = c("CAR", "B7C2", "B7"),
                        alpha = 0.05)

```

# Annotation Table

```{r}
sf.df.data.df = sf.df.data %>% 
  rownames_to_column(var = "gene_id")
sf.df.data.df
```

```{r}
deseqresult2df(res.sf.df.dds) %>% 
  dplyr::rename(gene_id = id) %>% 
  mutate(gene_symbol = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  mutate(gene_entrez = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first")) %>% 
  mutate(gene_desc = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "GENENAME", multiVals = "first")) %>% 
  relocate(gene_symbol, gene_entrez, gene_desc, .after = gene_id) %>% 
  left_join(sf.df.data.df, by = c("gene_id"="gene_id")) -> res.sf.df.dds.annotation
```

```{r}
View(res.sf.df.dds.annotation)
```

## Export Result

Open in Microsoft Excel ...

```{r}
write_csv(res.sf.df.dds.annotation, file = "CAR_DESeq2_Analysis_Result.csv")
```

# Quick Assessemnt

```{r}
sig.gene.cutoff = 0.000005 ## Reduce significant gene list to < 2,000

res.sf.df.dds.annotation %>% 
  ## Assess gene filtering:
  dplyr::filter(padj <= sig.gene.cutoff) 
```

# GeneTonic Prep

## GeneTonic Annotation

```{r}
deseqresult2df(il8.res.sf.df.dds) %>% 
  dplyr::rename(gene_id = id) %>% 
  mutate(gene_name = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  dplyr::select(gene_id, gene_name) -> sf.df.annotation
sf.df.annotation
```

## Enrichment Analysis

### Extract gene symbol

```{r}
deseqresult2df(res.sf.df.dds) %>% 
  # rownames_to_column(var = "Ensembl") %>% 
  mutate(Symbol = mapIds(org.Hs.eg.db, keys = id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  dplyr::filter(padj < sig.gene.cutoff) %>% 
  pull(Symbol) -> enrich.of.interest
length(enrich.of.interest)
head(enrich.of.interest)
```

## Background gene symbol

```{r}
deseqresult2df(res.sf.df.dds) %>% 
  mutate(Symbol = mapIds(org.Hs.eg.db, keys = id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  pull(Symbol) -> bg_ids
length(bg_ids)
```

## Perform Pathway Analysis

```{r}
topgo.res.sf.df.dds = pcaExplorer::topGOtable(enrich.of.interest,
                                              bg_ids,
                                              ontology = "BP",
                                              mapping = "org.Hs.eg.db",
                                              geneID = "symbol")

### Reshape for Genetonic
topgo.res.sf.df.dds = shake_topGOtableResult(topgo.res.sf.df.dds)
# rownames(topgo.res.sf.df.dds) = NULL
topgo.res.sf.df.dds = get_aggrscores(topgo.res.sf.df.dds, res.sf.df.dds, sf.df.annotation)
```

```{r}
topgo.res.sf.df.dds
```

# Genetonic

## SAVE Genetonic

```{r}
save(sf.df.dds, res.sf.df.dds, topgo.res.sf.df.dds, sf.df.annotation, file = "GeneTonic_inputs.Rdata")
```

```{r}
load("GeneTonic_inputs.Rdata")
```

Submit data objects to `Genetonic` web browser

```{r, eval=FALSE}
GeneTonic(dds = sf.df.dds,
          res_de = res.sf.df.dds,
          res_enrich = topgo.res.sf.df.dds,
          annotation_obj = sf.df.annotation,
          project_id = "TEST1")
```
