---
title: "2023-08-16_bulk_RNAseq_subset_analysis"
author: "Eric Hoffmeyer"
date: "2023-08-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load Libraries

These libraries are required to perform this RNAseq analysis. Make sure you have install them before begin the analysis.

```{r load libraries}
suppressPackageStartupMessages({
library(tidyverse)
library(tidybulk)
library(janitor)
library(plotly)
library(ggrepel)
library(DESeq2)
library(GeneTonic)
library(pcaExplorer)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(topGO)
})
```

# Data Dictionary

Creating a variable with the names of all the sample folders

```{r dir.names}
dir.names = c(
"quant_files/D1_B7C2_Il8_stim_S8_L002",
"quant_files/D1_B7C2_bead_stim_S14_L002",
"quant_files/D1_B7C2_no_stim_S2_L002",
"quant_files/D1_B7_Il8_stim_S7_L002",
"quant_files/D1_B7_bead_stim_S13_L002",
"quant_files/D1_B7_no_stim_S1_L002",
"quant_files/D2_B7C2_Il8_stim_S10_L002",
"quant_files/D2_B7C2_bead_stim_S16_L002",
"quant_files/D2_B7C2_no_stim_S4_L002",
"quant_files/D2_B7_Il8_stim_S9_L002",
"quant_files/D2_B7_bead_stim_S15_L002",
"quant_files/D2_B7_no_stim_S3_L002",
"quant_files/D3_B7C2_Il8_stim_S12_L002",
"quant_files/D3_B7C2_bead_stim_S18_L002",
"quant_files/D3_B7C2_no_stim_S6_L002",
"quant_files/D3_B7_Il8_stim_S11_L002",
"quant_files/D3_B7_bead_stim_S17_L002",
"quant_files/D3_B7_no_stim_S5_L002"
)
```

Convert the `vector` to `data.frame`.

```{r data dictionary start}
Data.dictionary <- data.frame(Orig.Name = dir.names, Folder.Name = dir.names) %>% 
  mutate(Folder.Name = str_remove(Folder.Name, pattern = "quant_files/"))
```

A `DataFrame`:

```{r view}
Data.dictionary
```

## Transform Dictionary

Perform these steps to transform the `data.frame` into desirable format.

```{r format data dictionary}
Data.dictionary <- Data.dictionary %>% 
  ## Create or duplicate the "Orig.Name" column and called it "Dup"
  mutate(Dup = Folder.Name) %>% 
  ## Breakup (separate) the "Dup" column using "_" as its separator
  separate_wider_delim(Dup, names = c("donor", "CAR", "stim", "line1", "line2", "line3"), delim = "_") %>%
  ## Create a new column called "col.name" by modifying "Orig.Name"
  ## by removing "_S(any digit)_L002"
  mutate(col.name = str_remove(Folder.Name, pattern = "_stim_S([:digit:]+)_L002")) %>%
  # Relocate the new "col.name" column right after the first column; "Org.Name"
  # relocate(col.name, .after = Orig.Name) %>% 
  ## Make donor, CAR, and stim columns a factors
  mutate(donor = as.factor(donor)) %>% 
  mutate(CAR = as.factor(CAR)) %>% 
  mutate(stim = as.factor(stim)) %>% 
  # mutate(col.name = str_c(donor, CAR, stim, sep = "_")) %>% 
  ## Remove line1, line2, line3 columns, as we don't need them.
  dplyr::select(-c("line1", "line2", "line3"))
```

```{r view}
Data.dictionary
```

# Transcript-Gene-Pair: `tx2gene_df`

We previously created the gtf, lookup table, and data matrix made from the salmon quant file loop. We will simply load it here.

## Load the saved copy

```{r load sf.df}
load(file = "data/sf.df.Rdata")
```

## Define subsets

Breaking up the data into three subsets: il8, bead, and bead (stim condition). DESeq will compare the B7C2 CAR to the B7 CAR in each of these stim conditions.

```{r subset data}
il8.subset.samples <- dplyr::filter(Data.dictionary, stim == "Il8")
il8.subset.samples <- il8.subset.samples$col.name

bead.subset.samples <- dplyr::filter(Data.dictionary, stim == "bead")
bead.subset.samples <- bead.subset.samples$col.name

no.subset.samples <- dplyr::filter(Data.dictionary, stim == "no")
no.subset.samples <- no.subset.samples$col.name
```

## Redefine meta data into subsets

Create data dictionaries that correspond to the subsets.

```{r subset data dictionaries}
il8.data.dictionary <- dplyr::filter(Data.dictionary, col.name %in% il8.subset.samples)

bead.data.dictionary <- dplyr::filter(Data.dictionary, col.name %in% bead.subset.samples)

no.data.dictionary <- dplyr::filter(Data.dictionary, col.name %in% no.subset.samples)
```

# Subset and filter count data

```{r filter count data}
il8.sf.df.data <- sf.df %>%
  dplyr::select(all_of(il8.subset.samples)) %>%
  drop_na() %>% 
  ## Filter low expressors (<10)
  mutate(Rowsum = apply(., 1, function(x) sum(x >= 5))) %>% 
  dplyr::filter(Rowsum == dim(il8.data.dictionary)[1]) %>% 
  ## no longer need the column "Rowsum", we can remove it
  dplyr::select(-Rowsum)

bead.sf.df.data <- sf.df %>%
  dplyr::select(all_of(bead.subset.samples)) %>%
  drop_na() %>% 
  ## Filter low expressors (<10)
  mutate(Rowsum = apply(., 1, function(x) sum(x >= 5))) %>% 
  dplyr::filter(Rowsum == dim(bead.data.dictionary)[1]) %>% 
  ## no longer need the column "Rowsum", we can remove it
  dplyr::select(-Rowsum)

no.sf.df.data <- sf.df %>%
  dplyr::select(all_of(no.subset.samples)) %>%
  drop_na() %>% 
  ## Filter low expressors (<10)
  mutate(Rowsum = apply(., 1, function(x) sum(x >= 4))) %>% 
  dplyr::filter(Rowsum == dim(no.data.dictionary)[1]) %>% 
  ## no longer need the column "Rowsum", we can remove it
  dplyr::select(-Rowsum)
```

# DESeq Object

## Create SummarizeExperiment Object

```{r Summarize Experiment}
il8.sf.df.se <- SummarizedExperiment(assays = list(counts = as.matrix(il8.sf.df.data)),
                                colData = il8.data.dictionary)

bead.sf.df.se <- SummarizedExperiment(assays = list(counts = as.matrix(bead.sf.df.data)),
                                colData = bead.data.dictionary)

no.sf.df.se <- SummarizedExperiment(assays = list(counts = as.matrix(no.sf.df.data)),
                                colData = no.data.dictionary)
```

## Create DESeq Object

The design will use CAR as the factor to run the comparison.

```{r DESeqDataSet}
il8.sf.df.dds <- DESeqDataSet(il8.sf.df.se, design = ~CAR)

bead.sf.df.dds <- DESeqDataSet(bead.sf.df.se, design = ~CAR)

no.sf.df.dds <- DESeqDataSet(no.sf.df.se, design = ~CAR)
```

# PCA Plots

```{r PCA}
il8.sf.df.se.pca <- il8.sf.df.se %>% 
  keep_abundant() %>% 
  scale_abundance() %>% 
  reduce_dimensions(method = "PCA", .dims = 3)
```

## 2D PCA
Set up color palette and theme

```{r color theme}
# Use colourblind-friendly colours
friendly_cols <- dittoSeq::dittoColors()

# Set theme
custom_theme <-
  list(
    scale_fill_manual(values = friendly_cols),
    scale_color_manual(values = friendly_cols),
    theme_bw() +
      theme(
        panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.major = element_line(size = 0.2),
        panel.grid.mibeadr = element_line(size = 0.1),
        text = element_text(size = 9),
        legend.position = "bottom",
        strip.background = element_blank(),
        axis.title.x = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.title.y = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1)
      )
  )

```

Plot PCA

```{r PCA plot}
il8.sf.df.se.pca %>% 
  pivot_sample() %>% 
  ggplot(aes(x = PC1, y = PC2, colour = donor)) +
  geom_point() +
  geom_text_repel(aes(label = .sample), show.legend = FALSE, size = 3) +
  custom_theme

```

# DE Analysis

## Statistical Analysis

```{r DESeq on subsets}
il8.sf.df.dds <- DESeq(il8.sf.df.dds)

bead.sf.df.dds <- DESeq(bead.sf.df.dds)

no.sf.df.dds <- DESeq(no.sf.df.dds)
```

## Analysis Result

```{r results()}
il8.res.sf.df.dds <- results(il8.sf.df.dds,
                        contrast = c("CAR", "B7C2", "B7"),
                        alpha = 0.05)

bead.res.sf.df.dds <- results(bead.sf.df.dds,
                        contrast = c("CAR", "B7C2", "B7"),
                        alpha = 0.05)

no.res.sf.df.dds <- results(no.sf.df.dds,
                        contrast = c("CAR", "B7C2", "B7"),
                        alpha = 0.05)
```

# annotation Table

```{r start annotation tables}
il8.sf.df.data.df <- il8.sf.df.data %>% 
  rownames_to_column(var = "gene_id")

bead.sf.df.data.df <- bead.sf.df.data %>% 
  rownames_to_column(var = "gene_id")

no.sf.df.data.df <- no.sf.df.data %>% 
  rownames_to_column(var = "gene_id")
```

```{r add gene_symbol, gene_entrez, and gene_desc to annotation tables}
il8.res.sf.df.dds.annotation <- deseqresult2df(il8.res.sf.df.dds) %>% 
  dplyr::rename(gene_id = id) %>% 
  mutate(gene_symbol = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  mutate(gene_entrez = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first")) %>% 
  mutate(gene_desc = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "GENENAME", multiVals = "first")) %>% 
  relocate(gene_symbol, gene_entrez, gene_desc, .after = gene_id) %>% 
  left_join(il8.sf.df.data.df, by = c("gene_id"="gene_id"))

bead.res.sf.df.dds.annotation <- deseqresult2df(bead.res.sf.df.dds) %>% 
  dplyr::rename(gene_id = id) %>% 
  mutate(gene_symbol = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  mutate(gene_entrez = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first")) %>% 
  mutate(gene_desc = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "GENENAME", multiVals = "first")) %>% 
  relocate(gene_symbol, gene_entrez, gene_desc, .after = gene_id) %>% 
  left_join(bead.sf.df.data.df, by = c("gene_id"="gene_id"))

no.res.sf.df.dds.annotation <- deseqresult2df(no.res.sf.df.dds) %>% 
  dplyr::rename(gene_id = id) %>% 
  mutate(gene_symbol = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  mutate(gene_entrez = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "ENTREZID", multiVals = "first")) %>% 
  mutate(gene_desc = mapIds(org.Hs.eg.db, keys = gene_id, keytype = "ENSEMBL", column = "GENENAME", multiVals = "first")) %>% 
  relocate(gene_symbol, gene_entrez, gene_desc, .after = gene_id) %>% 
  left_join(no.sf.df.data.df, by = c("gene_id"="gene_id"))
```

```{r view}
View(il8.res.sf.df.dds.annotation)
```

## Export Result

Write DESeq results to cxv files

```{r write DESeq results}
write_csv(il8.res.sf.df.dds.annotation, file = "deseq_results/il8_DESeq2_analysis.csv")

write_csv(bead.res.sf.df.dds.annotation, file = "deseq_results/bead_DESeq2_analysis.csv")

write_csv(no.res.sf.df.dds.annotation, file = "deseq_results/no_stim_DESeq2_anaylsis.csv")
```

# Quick Assessemnt

This DOES NOT change the dataframes. It is only meant to look at the number of significant genes to determine if we need to trim the lists down before further processing.

```{r filter padj <= 0.05}
sig.gene.cutoff = 0.05 ## Reduce significant gene list only p adjusted significant genes

il8.res.sf.df.dds.annotation %>% 
  dplyr::filter(padj <= sig.gene.cutoff) 

bead.res.sf.df.dds.annotation |>
  dplyr::filter(padj <= sig.gene.cutoff)

no.res.sf.df.dds.annotation |>
  dplyr::filter(padj <= sig.gene.cutoff)
```

# GeneTonic Prep

## GeneTonic annotation

```{r create/format annotation df for GeneTonic}
il8.sf.df.annotation <- deseqresult2df(il8.res.sf.df.dds) %>% 
  dplyr::rename(gene_id = id) %>% 
  mutate(gene_name = mapIds(org.Hs.eg.db, 
                            keys = gene_id, 
                            keytype = "ENSEMBL", 
                            column = "SYMBOL", 
                            multiVals = "first")) %>% 
  dplyr::select(gene_id, gene_name) 

bead.sf.df.annotation <- deseqresult2df(bead.res.sf.df.dds) %>% 
  dplyr::rename(gene_id = id) %>% 
  mutate(gene_name = mapIds(org.Hs.eg.db, 
                            keys = gene_id, 
                            keytype = "ENSEMBL", 
                            column = "SYMBOL", 
                            multiVals = "first")) %>% 
  dplyr::select(gene_id, gene_name)

no.sf.df.annotation <- deseqresult2df(no.res.sf.df.dds) %>% 
  dplyr::rename(gene_id = id) %>% 
  mutate(gene_name = mapIds(org.Hs.eg.db, 
                            keys = gene_id, 
                            keytype = "ENSEMBL", 
                            column = "SYMBOL", 
                            multiVals = "first")) %>% 
  dplyr::select(gene_id, gene_name)
```

## Enrichment Analysis

### Extract gene symbol

```{r enriched genes vectors for pathway analysis}
il8.enrich.of.interest <- deseqresult2df(il8.res.sf.df.dds) %>% 
  # rownames_to_column(var = "Ensembl") %>% 
  mutate(Symbol = mapIds(org.Hs.eg.db, keys = id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  dplyr::filter(padj < sig.gene.cutoff) %>% 
  pull(Symbol)

bead.enrich.of.interest <- deseqresult2df(bead.res.sf.df.dds) %>% 
  # rownames_to_column(var = "Ensembl") %>% 
  mutate(Symbol = mapIds(org.Hs.eg.db, keys = id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  dplyr::filter(padj < sig.gene.cutoff) %>% 
  pull(Symbol)

no.enrich.of.interest <- deseqresult2df(no.res.sf.df.dds) %>% 
  # rownames_to_column(var = "Ensembl") %>% 
  mutate(Symbol = mapIds(org.Hs.eg.db, keys = id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  dplyr::filter(padj < sig.gene.cutoff) %>% 
  pull(Symbol)

#length(enrich.of.interest)
#head(enrich.of.interest)
```

## Background gene symbol

```{r background gene vectors}
il8.bg.ids <- deseqresult2df(il8.res.sf.df.dds) %>% 
  mutate(Symbol = mapIds(org.Hs.eg.db, keys = id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  pull(Symbol)
#length(il8.bg.ids)

bead.bg.ids <- deseqresult2df(bead.res.sf.df.dds) %>% 
  mutate(Symbol = mapIds(org.Hs.eg.db, keys = id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  pull(Symbol)

no.bg.ids <- deseqresult2df(no.res.sf.df.dds) %>% 
  mutate(Symbol = mapIds(org.Hs.eg.db, keys = id, keytype = "ENSEMBL", column = "SYMBOL", multiVals = "first")) %>% 
  pull(Symbol)
```

## Perform Pathway Analysis

### IL-8 stim

Here is where I will deviate from Tzu's original script a little. While BP (Biological Processes) is the main gene ontology branch people consider, I want to run all three ontologies.

```{r il8 pathway analysis}
il8.topgo.res.sf.df.dds = pcaExplorer::topGOtable(il8.enrich.of.interest,
                                              il8.bg.ids,
                                              ontology = "BP",
                                              mapping = "org.Hs.eg.db",
                                              geneID = "symbol")

## Reshape for Genetonic
il8.topgo.res.sf.df.dds = shake_topGOtableResult(il8.topgo.res.sf.df.dds)
il8.topgo.res.sf.df.dds = get_aggrscores(il8.topgo.res.sf.df.dds, il8.res.sf.df.dds, il8.sf.df.annotation)

il8.mf.topgo.res.sf.df.dds = pcaExplorer::topGOtable(il8.enrich.of.interest,
                                              il8.bg.ids,
                                              ontology = "MF",
                                              mapping = "org.Hs.eg.db",
                                              geneID = "symbol")

## Reshape for Genetonic
il8.mf.topgo.res.sf.df.dds = shake_topGOtableResult(il8.mf.topgo.res.sf.df.dds)
il8.mf.topgo.res.sf.df.dds = get_aggrscores(il8.mf.topgo.res.sf.df.dds, il8.res.sf.df.dds, il8.sf.df.annotation)

il8.cc.topgo.res.sf.df.dds = pcaExplorer::topGOtable(il8.enrich.of.interest,
                                              il8.bg.ids,
                                              ontology = "CC",
                                              mapping = "org.Hs.eg.db",
                                              geneID = "symbol")

## Reshape for Genetonic
il8.cc.topgo.res.sf.df.dds = shake_topGOtableResult(il8.cc.topgo.res.sf.df.dds)
il8.cc.topgo.res.sf.df.dds = get_aggrscores(il8.cc.topgo.res.sf.df.dds, il8.res.sf.df.dds, il8.sf.df.annotation)
```

Now I can bind these resulting dataframes together and arrange them by p value. But the combined dataframe is too large to interact with some GeneTonic functions, so I will trim it down to the typical 200 rows.

```{r combine and trim analysis results}
il8.topgo.res.sf.df.dds <- rbind(il8.topgo.res.sf.df.dds, il8.mf.topgo.res.sf.df.dds, il8.cc.topgo.res.sf.df.dds) |>
  arrange(gs_pvalue)

il8.topgo.res.sf.df.dds <- il8.topgo.res.sf.df.dds[1:200,]
```

```{r save dds files for genetonic}
save(il8.sf.df.dds, il8.res.sf.df.dds, il8.topgo.res.sf.df.dds, il8.sf.df.annotation,
     file = "genetonic_results/il8_GeneTonic_inputs.Rdata")
```

Submit data objects to `Genetonic` web browser

```{r il8 genetonic, eval=FALSE}
GeneTonic(dds = il8.sf.df.dds,
          res_de = il8.res.sf.df.dds,
          res_enrich = il8.topgo.res.sf.df.dds,
          annotation_obj = il8.sf.df.annotation,
          project_id = "il8 stim")
```

Remove all il8 variables
```{r}
rm.list <- grep("il8", x = ls(), value = T)

rm(list = rm.list)
```

### bead stim

Repeating the process for bead stim condition.

```{r bead pathway analysis}
bead.topgo.res.sf.df.dds = pcaExplorer::topGOtable(bead.enrich.of.interest,
                                              bead.bg.ids,
                                              ontology = "BP",
                                              mapping = "org.Hs.eg.db",
                                              geneID = "symbol")

## Reshape for Genetonic
bead.topgo.res.sf.df.dds = shake_topGOtableResult(bead.topgo.res.sf.df.dds)
bead.topgo.res.sf.df.dds = get_aggrscores(bead.topgo.res.sf.df.dds, bead.res.sf.df.dds, bead.sf.df.annotation)

bead.mf.topgo.res.sf.df.dds = pcaExplorer::topGOtable(bead.enrich.of.interest,
                                              bead.bg.ids,
                                              ontology = "MF",
                                              mapping = "org.Hs.eg.db",
                                              geneID = "symbol")

## Reshape for Genetonic
bead.mf.topgo.res.sf.df.dds = shake_topGOtableResult(bead.mf.topgo.res.sf.df.dds)
bead.mf.topgo.res.sf.df.dds = get_aggrscores(bead.mf.topgo.res.sf.df.dds, bead.res.sf.df.dds, bead.sf.df.annotation)

bead.cc.topgo.res.sf.df.dds = pcaExplorer::topGOtable(bead.enrich.of.interest,
                                              bead.bg.ids,
                                              ontology = "CC",
                                              mapping = "org.Hs.eg.db",
                                              geneID = "symbol")

## Reshape for Genetonic
bead.cc.topgo.res.sf.df.dds = shake_topGOtableResult(bead.cc.topgo.res.sf.df.dds)
bead.cc.topgo.res.sf.df.dds = get_aggrscores(bead.cc.topgo.res.sf.df.dds, bead.res.sf.df.dds, bead.sf.df.annotation)
```


```{r combine and trim analysis results}
bead.topgo.res.sf.df.dds <- rbind(bead.topgo.res.sf.df.dds, bead.mf.topgo.res.sf.df.dds, bead.cc.topgo.res.sf.df.dds) |>
  arrange(gs_pvalue)

bead.topgo.res.sf.df.dds <- bead.topgo.res.sf.df.dds[1:200,]
```

```{r save dds files for genetonic}
save(bead.sf.df.dds, bead.res.sf.df.dds, bead.topgo.res.sf.df.dds, bead.sf.df.annotation,
     file = "genetonic_results/bead_GeneTonic_inputs.Rdata")
```

Submit data objects to `Genetonic` web browser

```{r bead genetonic, eval=FALSE}
GeneTonic(dds = bead.sf.df.dds,
          res_de = bead.res.sf.df.dds,
          res_enrich = bead.topgo.res.sf.df.dds,
          annotation_obj = bead.sf.df.annotation,
          project_id = "bead stim")
```

Remove all bead variables
```{r}
rm.list <- grep("bead", x = ls(), value = T)

rm(list = rm.list)
```


### no stim

Repeating the process for no stim condition.

```{r no pathway analysis}
no.topgo.res.sf.df.dds = pcaExplorer::topGOtable(no.enrich.of.interest,
                                              no.bg.ids,
                                              ontology = "BP",
                                              mapping = "org.Hs.eg.db",
                                              geneID = "symbol")

## Reshape for Genetonic
no.topgo.res.sf.df.dds = shake_topGOtableResult(no.topgo.res.sf.df.dds)
no.topgo.res.sf.df.dds = get_aggrscores(no.topgo.res.sf.df.dds, no.res.sf.df.dds, no.sf.df.annotation)

no.mf.topgo.res.sf.df.dds = pcaExplorer::topGOtable(no.enrich.of.interest,
                                              no.bg.ids,
                                              ontology = "MF",
                                              mapping = "org.Hs.eg.db",
                                              geneID = "symbol")

## Reshape for Genetonic
no.mf.topgo.res.sf.df.dds = shake_topGOtableResult(no.mf.topgo.res.sf.df.dds)
no.mf.topgo.res.sf.df.dds = get_aggrscores(no.mf.topgo.res.sf.df.dds, no.res.sf.df.dds, no.sf.df.annotation)

no.cc.topgo.res.sf.df.dds = pcaExplorer::topGOtable(no.enrich.of.interest,
                                              no.bg.ids,
                                              ontology = "CC",
                                              mapping = "org.Hs.eg.db",
                                              geneID = "symbol")

## Reshape for Genetonic
no.cc.topgo.res.sf.df.dds = shake_topGOtableResult(no.cc.topgo.res.sf.df.dds)
no.cc.topgo.res.sf.df.dds = get_aggrscores(no.cc.topgo.res.sf.df.dds, no.res.sf.df.dds, no.sf.df.annotation)
```


```{r combine and trim analysis results}
no.topgo.res.sf.df.dds <- rbind(no.topgo.res.sf.df.dds, no.mf.topgo.res.sf.df.dds, no.cc.topgo.res.sf.df.dds) |>
  arrange(gs_pvalue)

no.topgo.res.sf.df.dds <- no.topgo.res.sf.df.dds[1:200,]
```

```{r save dds files for genetonic}
save(no.sf.df.dds, no.res.sf.df.dds, no.topgo.res.sf.df.dds, no.sf.df.annotation,
     file = "genetonic_results/no_GeneTonic_inputs.Rdata")
```

Submit data objects to `Genetonic` web browser

```{r no genetonic, eval=FALSE}
GeneTonic(dds = no.sf.df.dds,
          res_de = no.res.sf.df.dds,
          res_enrich = no.topgo.res.sf.df.dds,
          annotation_obj = no.sf.df.annotation,
          project_id = "no stim")
```

Remove all no variables
```{r}
rm.list <- grep("no", x = ls(), value = T)

rm(list = rm.list)
```





# Session Info
```{r}
sessionInfo()
```

